### ISSUE TITLE
Numpy interface corrupts array argument

### ISSUE THREAD

**Issue by bwohlberg** [pyFFTW/pyFFTW#197]
There appears to be a platform-dependent bug in the numpy interface that results in corruption of the array passed to ``pyfftw.interfaces.numpy_fft.rfftn``. Under 32 bit Ubuntu 16.04 running on an Intel T2500 CPU, the following script

```
from __future__ import print_function

import hashlib
import numpy as np
import pyfftw

pyfftw.interfaces.cache.enable()

def rfftn(a, s=None, axes=None):
    return pyfftw.interfaces.numpy_fft.rfftn(a, s=s, axes=axes,
                overwrite_input=False, planner_effort='FFTW_MEASURE')

def hash(x):
    return hashlib.md5(x).hexdigest()[0:8]

np.random.seed(12345)
N = 64
U = np.random.randn(N, N).astype(np.float32)
V = np.random.randn(N, N).astype(np.float32)

for k in range(3):
    print(hash(U), hash(V))
    x = rfftn(U, None, axes=(0,1))
    y = rfftn(V, None, axes=(0,1))
```

gives output

```
a7662abd bb54ccca
a7662abd bb54ccca
a7662abd a7662abd
```

or

```
a7662abd bb54ccca
bb54ccca bb54ccca
bb54ccca bb54ccca
```

indicating that calling ``pyfftw.interfaces.numpy_fft.rfftn`` with ``U`` and then ``V`` results in ``U`` overwriting ``V`` or vice-versa.

This only seems to occur when the script is run with Python 3.5 (``python3`` on Ubuntu 16.04), and does not occur when the initial call to ``pyfftw.interfaces.cache.enable()`` is removed.

**Comment by hgomersall** [comment]
I can't replicate this on my machine (16.04 Ubuntu but 64-bit) (I just get the . Are you able to pin down the bug better? It is probably happening in [pyfftw.interfaces._utils](https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/interfaces/_utils.py), with arrays not being protected properly.

Thanks!

**Comment by bwohlberg** [comment]
From the attached debugging code (a modified version of the original
bug demonstration script, a version of pyfftw.interfaces._utils with
added debugging code, and the output of the script), it appears that
the corruption is occurring in the final "FFTW_object(input_array=a,
...)" call in _utils.py, and affects the "other" array, i.e., the
array to which rfftn was previously applied, not the argument a
itself.

Does that help?


On Mon, Nov 20, 2017 at 5:01 AM, Henry Gomersall
<notifications@github.com> wrote:
> I can't replicate this on my machine (16.04 Ubuntu but 64-bit) (I just get
> the . Are you able to pin down the bug better? It is probably happening in
> pyfftw.interfaces._utils, with arrays not being protected properly.
>
> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

$ python3 bug_demo.py
before rfftn(U): a7662abd bb54ccca
A1 a7662abd bb54ccca
A2 a7662abd bb54ccca
A3 a7662abd bb54ccca
after rfftn(U), before  rfftn(V): a7662abd bb54ccca
A1 bb54ccca a7662abd
A4 bb54ccca a7662abd
A5 bb54ccca a7662abd
A6 bb54ccca a7662abd
after rfftn(V): a7662abd bb54ccca
before rfftn(U): a7662abd bb54ccca
A1 a7662abd bb54ccca
A4 a7662abd bb54ccca
A5 a7662abd bb54ccca
A6 a7662abd a7662abd
after rfftn(U), before  rfftn(V): a7662abd a7662abd
A1 a7662abd a7662abd
A4 a7662abd a7662abd
A5 a7662abd a7662abd
A6 a7662abd a7662abd
after rfftn(V): a7662abd a7662abd
before rfftn(U): a7662abd a7662abd
A1 a7662abd a7662abd
A4 a7662abd a7662abd
A5 a7662abd a7662abd
A6 a7662abd a7662abd
after rfftn(U), before  rfftn(V): a7662abd a7662abd
A1 a7662abd a7662abd
A4 a7662abd a7662abd
A5 a7662abd a7662abd
A6 a7662abd a7662abd
after rfftn(V): a7662abd a7662abd

**Comment by hgomersall** [comment]
No, I don't have the email attachment. Can you confirm that the logic is correct? `rfftn` ought to not clobber the input array in calls to FFTW, so I'm curious why it should be. I'd be interested to know, if you can dig a little deeper, where _exactly_ the problem is occurring.

Are you able to build more and see whether the corruption is occurring in the call to FFTW itself? It should be sufficient to inspect the output around [here](https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/pyfftw.pyx#L1638), but the entry point to the FFTW object is through the [__call__ method](https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/pyfftw.pyx#L1369) so that would be useful to explore too.

**Comment by bwohlberg** [comment]
I'll include the missing attachments in a response directly after this one. I should emphasise that ``rfftn`` is not directly clobbering *its own* input array, it's overwriting the input array to a *previous* call to ``rfftn`` with the current input array. I assume that this is related to the way the cache mechanism handles the input array argument, since it doesn't occur if the cache is disabled. Given this point, perhaps you would suggest somewhere else in the code to continue the investigation?

**Comment by bwohlberg** [comment]
modified bug demonstration script:
```
from __future__ import print_function

import hashlib
import numpy as np
import pyfftw


pyfftw.interfaces.cache.enable()

def rfftn(a, s=None, axes=None):
    return pyfftw.interfaces.numpy_fft.rfftn(a, s=s, axes=axes,
                overwrite_input=False, planner_effort='FFTW_MEASURE')


def hash(x):
    return hashlib.md5(x).hexdigest()[0:8]


np.random.seed(12345)
N = 64
U = np.random.randn(N, N).astype(np.float32)
V = np.random.randn(N, N).astype(np.float32)

for k in range(3):
    print('before rfftn(U):', hash(U), hash(V))
    pyfftw.interfaces._utils.other_array = V
    x = rfftn(U, None, axes=(0,1))
    print('after rfftn(U), before  rfftn(V):', hash(U), hash(V))
    pyfftw.interfaces._utils.other_array = U
    y = rfftn(V, None, axes=(0,1))
    print('after rfftn(V):', hash(U), hash(V))
```

modified pyfftw.interfaces._utils (header comment omitted):
```
'''
Utility functions for the interfaces routines
'''

import hashlib

def hash(x):
    return hashlib.md5(x).hexdigest()[0:8]

import pyfftw.builders as builders
import pyfftw
import numpy
from . import cache

other_array = None

def _Xfftn(a, s, axes, overwrite_input, planner_effort,
        threads, auto_align_input, auto_contiguous,
        calling_func, normalise_idft=True, ortho=False):


    global other_array

    work_with_copy = False

    print('A1', hash(a), hash(other_array))

    a = numpy.asanyarray(a)

    try:
        s = tuple(s)
    except TypeError:
        pass

    try:
        axes = tuple(axes)
    except TypeError:
        pass

    if calling_func in ('irfft2', 'irfftn'):
        # overwrite_input is not an argument to irfft2 or irfftn
        args = (planner_effort, threads, auto_align_input, auto_contiguous)

        if not overwrite_input:
            # Only irfft2 and irfftn have overwriting the input
            # as the default (and so require the input array to
            # be reloaded).
            work_with_copy = True
    else:
        args = (overwrite_input, planner_effort, threads,
                auto_align_input, auto_contiguous)

        if not a.flags.writeable:
            # Special case of a locked array - always work with a
            # copy.  See issue #92.
            work_with_copy = True

            if overwrite_input:
                raise ValueError('overwrite_input cannot be True when the ' +
                                 'input array flags.writeable is False')

    if work_with_copy:
        # We make the copy before registering the key so that the
        # copy's stride information will be cached since this will be
        # used for planning.  Make sure the copy is byte aligned to
        # prevent further copying
        a_original = a
        a = pyfftw.empty_aligned(shape=a.shape, dtype=a.dtype)
        a[...] = a_original

    if cache.is_enabled():
        key = (calling_func, a.shape, a.strides, a.dtype, s.__hash__(),
               axes.__hash__(), args)

        try:
            if key in cache._fftw_cache:
                FFTW_object = cache._fftw_cache.lookup(key)
            else:
                FFTW_object = None

        except KeyError:
            # This occurs if the object has fallen out of the cache between
            # the check and the lookup
            FFTW_object = None

    if not cache.is_enabled() or FFTW_object is None:

        # If we're going to create a new FFTW object and are not
        # working with a copy, then we need to copy the input array to
        # preserve it, otherwise we can't actually  take the transform
        # of the input array! (in general, we have to assume that the
        # input array will be destroyed during planning).
        if not work_with_copy:
            a_copy = a.copy()

        planner_args = (a, s, axes) + args

        FFTW_object = getattr(builders, calling_func)(*planner_args)

        # Only copy if the input array is what was actually used
        # (otherwise it shouldn't be overwritten)
        if not work_with_copy and FFTW_object.input_array is a:
            a[:] = a_copy

        print('A2', hash(a), hash(other_array))

        if cache.is_enabled():
            cache._fftw_cache.insert(FFTW_object, key)

        output_array = FFTW_object(normalise_idft=normalise_idft, ortho=ortho)

        print('A3', hash(a), hash(other_array))
    else:

        print('A4', hash(a), hash(other_array))

        orig_output_array = FFTW_object.output_array
        output_shape = orig_output_array.shape
        output_dtype = orig_output_array.dtype
        output_alignment = FFTW_object.output_alignment

        output_array = pyfftw.empty_aligned(
            output_shape, output_dtype, n=output_alignment)

        print('A5', hash(a), hash(other_array))

        FFTW_object(input_array=a, output_array=output_array,
                normalise_idft=normalise_idft, ortho=ortho)

        print('A6', hash(a), hash(other_array))

    return output_array
```

**Comment by hgomersall** [comment]
I think the first thing is to establish it _is_ using the cache. You can check this by printing out the `id()` of the `FFTW_object` in `_utils.py` just before it's called.

Then check the address (e.g. `FFTW_object.input_array.ctypes.data`). These _should_ be different to the actual input addresses. I've noticed a plausible reason for the problem (though why it's not universal, I'm not sure).

**Comment by bwohlberg** [comment]
Here's the debugging output including the id and address (truncated after the problem becomes apparent):

> before rfftn(U): a7662abd bb54ccca
A1 a7662abd bb54ccca
A2 a7662abd bb54ccca
Before FFTW_object call ID:  3055638764
FFTW_object.input_array Address:  170174560   a Address:  170174560
After FFTW_object call ID:  3055638764
FFTW_object.input_array Address:  170174560   a Address:  170174560
A3 a7662abd bb54ccca
after rfftn(U), before  rfftn(V): a7662abd bb54ccca
A1 bb54ccca a7662abd
A4 bb54ccca a7662abd
A5 bb54ccca a7662abd
Before FFTW_object call ID:  3055638764
FFTW_object.input_array Address:  170174560   a Address:  170190952
After FFTW_object call ID:  3055638764
FFTW_object.input_array Address:  170174560   a Address:  170190952
A6 bb54ccca bb54ccca
after rfftn(V): bb54ccca bb54ccca

After the second  ``FFTW_object`` call, both arrays have the same content. I assume that the input array address on this second call should be the same as the ``a`` address, and not the same as the address of the array from the first ``FFTW_object`` call.

The modified version of ``_utils.py`` is included below:

```
'''
Utility functions for the interfaces routines
'''

import hashlib

def hash(x):
    return hashlib.md5(x).hexdigest()[0:8]

import pyfftw.builders as builders
import pyfftw
import numpy
from . import cache

other_array = None

def _Xfftn(a, s, axes, overwrite_input, planner_effort,
        threads, auto_align_input, auto_contiguous,
        calling_func, normalise_idft=True, ortho=False):


    global other_array

    work_with_copy = False

    print('A1', hash(a), hash(other_array))

    a = numpy.asanyarray(a)

    try:
        s = tuple(s)
    except TypeError:
        pass

    try:
        axes = tuple(axes)
    except TypeError:
        pass

    if calling_func in ('irfft2', 'irfftn'):
        # overwrite_input is not an argument to irfft2 or irfftn
        args = (planner_effort, threads, auto_align_input, auto_contiguous)

        if not overwrite_input:
            # Only irfft2 and irfftn have overwriting the input
            # as the default (and so require the input array to
            # be reloaded).
            work_with_copy = True
    else:
        args = (overwrite_input, planner_effort, threads,
                auto_align_input, auto_contiguous)

        if not a.flags.writeable:
            # Special case of a locked array - always work with a
            # copy.  See issue #92.
            work_with_copy = True

            if overwrite_input:
                raise ValueError('overwrite_input cannot be True when the ' +
                                 'input array flags.writeable is False')

    if work_with_copy:
        # We make the copy before registering the key so that the
        # copy's stride information will be cached since this will be
        # used for planning.  Make sure the copy is byte aligned to
        # prevent further copying
        a_original = a
        a = pyfftw.empty_aligned(shape=a.shape, dtype=a.dtype)
        a[...] = a_original

    if cache.is_enabled():
        key = (calling_func, a.shape, a.strides, a.dtype, s.__hash__(),
               axes.__hash__(), args)

        try:
            if key in cache._fftw_cache:
                FFTW_object = cache._fftw_cache.lookup(key)
            else:
                FFTW_object = None

        except KeyError:
            # This occurs if the object has fallen out of the cache between
            # the check and the lookup
            FFTW_object = None

    if not cache.is_enabled() or FFTW_object is None:

        # If we're going to create a new FFTW object and are not
        # working with a copy, then we need to copy the input array to
        # preserve it, otherwise we can't actually  take the transform
        # of the input array! (in general, we have to assume that the
        # input array will be destroyed during planning).
        if not work_with_copy:
            a_copy = a.copy()

        planner_args = (a, s, axes) + args

        FFTW_object = getattr(builders, calling_func)(*planner_args)

        # Only copy if the input array is what was actually used
        # (otherwise it shouldn't be overwritten)
        if not work_with_copy and FFTW_object.input_array is a:
            a[:] = a_copy

        print('A2', hash(a), hash(other_array))

        if cache.is_enabled():
            cache._fftw_cache.insert(FFTW_object, key)

        print('Before FFTW_object call', 'ID: ', id(FFTW_object))
        print('FFTW_object.input_array Address: ',
              FFTW_object.input_array.ctypes.data, '  a Address: ', a.ctypes.data)
        output_array = FFTW_object(normalise_idft=normalise_idft, ortho=ortho)
        print('After FFTW_object call', 'ID: ', id(FFTW_object))
        print('FFTW_object.input_array Address: ',
              FFTW_object.input_array.ctypes.data, '  a Address: ', a.ctypes.data)

        print('A3', hash(a), hash(other_array))
    else:

        print('A4', hash(a), hash(other_array))

        orig_output_array = FFTW_object.output_array
        output_shape = orig_output_array.shape
        output_dtype = orig_output_array.dtype
        output_alignment = FFTW_object.output_alignment

        output_array = pyfftw.empty_aligned(
            output_shape, output_dtype, n=output_alignment)

        print('A5', hash(a), hash(other_array))

        print('Before FFTW_object call', 'ID: ', id(FFTW_object))
        print('FFTW_object.input_array Address: ',
              FFTW_object.input_array.ctypes.data, '  a Address: ', a.ctypes.data)
        FFTW_object(input_array=a, output_array=output_array,
                normalise_idft=normalise_idft, ortho=ortho)
        print('After FFTW_object call', 'ID: ', id(FFTW_object))
        print('FFTW_object.input_array Address: ',
              FFTW_object.input_array.ctypes.data, '  a Address: ', a.ctypes.data)

        print('A6', hash(a), hash(other_array))

    return output_array
```

**Comment by hgomersall** [comment]
Hmmm, my thinking on this is that we have an alignment issue. The cache is returning an object that it thinks doesn't clobber the input because it will always update it, however the passed in second array has a different alignment to the original array (I suspect, no longer aligned), so the FFTW object is forced to copy it in and it does so over the top of the internal array (which is the correct behaviour). The problem is, the internal array is the same as the input array in the first instance.

I can't see what the obvious solution here is that doesn't result in always creating a copy. I think probably the cache should use the alignment as one of it's key parameters, but I'm not totally sure.

It would be interesting if you could investigate this for me. The alignment can be checked by looking at the address modulo some base alignment (say, 32).

**Comment by bwohlberg** [comment]
Your alignment hypothesis seems to be correct. Modifying the previous debugging code to print the address modulo 16 (the simd_alignment value for the machine on which it was run) gives

> before rfftn(U): bb54ccca ec339b61
A1 bb54ccca ec339b61
A2 bb54ccca ec339b61
Before FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  8
After FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  8
A3 bb54ccca ec339b61
after rfftn(U), before  rfftn(V): bb54ccca ec339b61
A1 ec339b61 bb54ccca
A4 ec339b61 bb54ccca
A5 ec339b61 bb54ccca
Before FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  0
After FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  0
A6 ec339b61 bb54ccca
after rfftn(V): bb54ccca ec339b61
>
> before rfftn(U): bb54ccca ec339b61
A1 bb54ccca ec339b61
A4 bb54ccca ec339b61
A5 bb54ccca ec339b61
Before FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  8
After FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  8
A6 bb54ccca bb54ccca
after rfftn(U), before  rfftn(V): bb54ccca bb54ccca
A1 bb54ccca bb54ccca
A4 bb54ccca bb54ccca
A5 bb54ccca bb54ccca
Before FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  0
After FFTW_object call ID:  3055286508
FFTW_object.input_array Address:  0   a Address:  0
A6 bb54ccca bb54ccca
after rfftn(V): bb54ccca bb54ccca

If the driver script is modified to ensure that the input arrays are correctly aligned, the array corruption bug disappears.

**Comment by hgomersall** [comment]
Perhaps the solution then is to enforce alignment during construction, and then enforce that with copies in `_utils.py`? This would at least limit the copies to the cases when there was no alignment (and the docs could point this out).

That seems to me to be a better solution than always copying...

**Comment by bwohlberg** [comment]
That seems reasonable, but your previous suggestion of including the alignment in the cache key also seems worth considering.

**Comment by hgomersall** [comment]
@bwohlberg hmmm I thought that wouldn't work for another reason but I'm not sure now...

Are you able to write a test that exercises this particular issue for me? That would be really helpful so we can sort this properly. Adding a key for alignment is easy so that would be a quick fix. Adding it to `test_pyfftw_numpy_interface.InterfacesNumpyFFTTestFFT` would seem appropriate. The alignment stuff can be forced by slicing out a suitably sliced array for an oversized array.

**Comment by bwohlberg** [comment]
My attempts at building ``pyfftw`` from source fail with
> KeyError: 'HAVE_DOUBLE_THREADS'

so I'm not able to add the test to the repo and handle your request via a pull request.

Below is a short ``pytest`` test that properly exercises the issue. I would expect it to be quite easy to convert to the ``unittest`` framework that you use, but let me know if you need help (or if you can suggest what the above error means, I can try to do it myself).

``` 
from __future__ import division
from builtins import object

import pytest
import hashlib

import numpy as np
import pyfftw


def rfftn(a, s=None, axes=None):
    return pyfftw.interfaces.numpy_fft.rfftn(a, s=s, axes=axes,
                overwrite_input=False, planner_effort='FFTW_MEASURE')


def hash(x):
    return hashlib.md5(x).hexdigest()


class TestSet01(object):

    def setup_method(self, method):
        N = 64
        pyfftw.interfaces.cache.enable()
        np.random.seed(12345)
        self.U = np.random.randn(N, N).astype(np.float32)
        self.V = np.random.randn(N, N).astype(np.float32)


    def test_01(self):
        # Test for pyfftw bug (on certain architectures) that corrupts the
        # content of argument variables. In this test, the bug copies the
        # content of V into U or vice-versa.
        U = self.U
        V = self.V
        uh = hash(U)
        vh = hash(V)
        x = rfftn(U, None, axes=(0,1))
        y = rfftn(V, None, axes=(0,1))
        assert(uh == hash(U))
        assert(vh == hash(V))
        x = rfftn(U, None, axes=(0,1))
        y = rfftn(V, None, axes=(0,1))
        assert(uh == hash(U))
        assert(vh == hash(V))
```

**Comment by hgomersall** [comment]
Interesting... Is that test reliable? It doesn't seem to explicitly set the alignment so I'd expect it to sometimes pass...

**Comment by bwohlberg** [comment]
Good point. This test was written before we discovered the alignment connection, but yes, it does seem to reliably fail when the issue is present. This is probably why two sets of ``rfttn`` calls were needed; if I were to redesign the test with the knowledge of the role of alignment, it would probably be possible to reduce it to a single set of ``rfftn`` calls.

**Comment by hgomersall** [comment]
That would be great. I can work with your style of test and incorporate that into the suite. Thanks!

**Comment by bwohlberg** [comment]
See below for the updated test. This one also fails on Ubuntu 16.04 64 bit; either the careful setting of the mis-alignment has removed the system-dependence of the problem, or there's a bug in the test that I'm overlooking.

```
from __future__ import division
from builtins import object

import pytest
import hashlib

import numpy as np
import pyfftw


def rfftn(a, s=None, axes=None):
    return pyfftw.interfaces.numpy_fft.rfftn(a, s=s, axes=axes,
                overwrite_input=False, planner_effort='FFTW_MEASURE')


def hash(x):
    return hashlib.md5(x).hexdigest()


# We test for pyfftw bugs here to help determine whether the failure of
# other tests is due to bugs in sporco or in pyffyw
class TestSet01(object):

    def setup_method(self, method):
        N = 64
        pyfftw.interfaces.cache.enable()
        np.random.seed(12345)
        Um = pyfftw.empty_aligned((N, N+1), dtype=np.float32, order='C')
        Vm = pyfftw.empty_aligned((N, N+1), dtype=np.float32, order='C')
        self.U = np.ndarray((N, N), dtype=Um.dtype, buffer=Um.data, offset=0)
        self.V = np.ndarray((N, N), dtype=Vm.dtype, buffer=Vm.data, offset=4)
        self.U[:] = np.random.randn(N, N).astype(np.float32)
        self.V[:] = np.random.randn(N, N).astype(np.float32)


    def test_01(self):
        # Test for pyfftw bug (on certain architectures) that corrupts the
        # content of argument variables. In this test, the bug copies the
        # content of V into U or vice-versa.
        U = self.U
        V = self.V
        uh = hash(U)
        vh = hash(V)
        x = rfftn(U, None, axes=(0,1))
        y = rfftn(V, None, axes=(0,1))
        assert(uh == hash(U))
        assert(vh == hash(V))
```

**Comment by hgomersall** [comment]
Perfect, thanks. I would expect it to fail on 64-bit, so I'm pleased it does :)

It might be a couple of days before I'm able to visit this. I don't think I'll forget, but if I do, please poke me.

**Comment by grlee77** [comment]
> My attempts at building pyfftw from source fail with
> KeyError: 'HAVE_DOUBLE_THREADS'

Can you verify whether or not the small change to `setup.py` proposed in #198 fixes this issue for you?  We recently started having failures with this same issue on the CI services.

**Comment by bwohlberg** [comment]
@grlee77 That fixed the build problems for me.

**Comment by hgomersall** [comment]
hopefully this is fixed in #199.

**Comment by hgomersall** [comment]
Fixed in #199.
